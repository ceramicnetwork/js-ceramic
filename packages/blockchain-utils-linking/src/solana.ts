import { AccountId } from 'caip'
import { AuthProvider } from './auth-provider.js'
import { asOldCaipString, getConsentMessage, LinkProof } from './util.js'
import * as uint8arrays from 'uint8arrays'
import * as sha256 from '@stablelib/sha256'

export const SOLANA_TESTNET_CHAIN_REF = '4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z' // Solana testnet
export const SOLANA_DEVNET_CHAIN_REF = 'EtWTRABZaYq6iMfeYKouRu166VU2xqa1' // Solana devnet
export const SOLANA_MAINNET_CHAIN_REF = '5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp' // Solana mainnet beta

export class SolanaAuthProvider implements AuthProvider {
  readonly isAuthProvider = true

  constructor(
    private readonly provider: any,
    private readonly address: string,
    private readonly chainRef: string
  ) {
    console.warn(
      'WARN: SolanaAuthProvider is not fully supported. You may encounter issues using this.'
    )
  }

  async accountId(): Promise<AccountId> {
    return new AccountId({
      address: this.address,
      chainId: `solana:${this.chainRef}`,
    })
  }

  async authenticate(message: string): Promise<string> {
    if (!this.provider.signMessage) {
      throw new Error(`Unsupported provider; provider must implement signMessage`)
    }
    const signatureBytes = await this.provider.signMessage(uint8arrays.fromString(message))
    const digest = sha256.hash(signatureBytes)
    return `0x${uint8arrays.toString(digest, 'base16')}`
  }

  async createLink(did: string): Promise<LinkProof> {
    if (!this.provider.signMessage) {
      throw new Error(`Unsupported provider; provider must implement signMessage`)
    }
    const { message, timestamp } = getConsentMessage(did, true)
    const accountID = await this.accountId()
    const signatureBytes = await this.provider.signMessage(uint8arrays.fromString(message))
    const signature = uint8arrays.toString(signatureBytes, 'base64')
    return {
      version: 2,
      type: 'solana',
      message,
      signature,
      account: asOldCaipString(accountID),
      timestamp,
    }
  }

  withAddress(address: string): AuthProvider {
    return new SolanaAuthProvider(this.provider, address, this.chainRef)
  }
}
