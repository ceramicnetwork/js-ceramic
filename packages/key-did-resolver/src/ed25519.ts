import { toString } from 'uint8arrays/to-string'
import { convertPublicKeyToX25519 } from '@stablelib/ed25519'

function encodeKey(key: Uint8Array): string {
  const bytes = new Uint8Array(key.length + 2)
  bytes[0] = 0xec
  // The multicodec is encoded as a varint so we need to add this.
  // See js-multicodec for a general implementation
  bytes[1] = 0x01
  bytes.set(key, 2)
  return `z${toString(bytes, 'base58btc')}`
}

/**
 * Constructs the document based on the method key
 */
export function keyToDidDoc(pubKeyBytes: Uint8Array, fingerprint: string): any {
  const did = `did:key:${fingerprint}`
  const keyId = `${did}#${fingerprint}`
  const x25519PubBytes = convertPublicKeyToX25519(pubKeyBytes)
  const x25519KeyId = `${did}#${encodeKey(x25519PubBytes)}`
  return {
    id: did,
    verificationMethod: [
      {
        id: keyId,
        type: 'Ed25519VerificationKey2018',
        controller: did,
        publicKeyBase58: toString(pubKeyBytes, 'base58btc'),
      },
    ],
    authentication: [keyId],
    assertionMethod: [keyId],
    capabilityDelegation: [keyId],
    capabilityInvocation: [keyId],
    keyAgreement: [
      {
        id: x25519KeyId,
        type: 'X25519KeyAgreementKey2019',
        controller: did,
        publicKeyBase58: toString(x25519PubBytes, 'base58btc'),
      },
    ],
  }
}
