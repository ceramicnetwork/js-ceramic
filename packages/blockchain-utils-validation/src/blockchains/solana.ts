import { BlockchainHandler } from '../blockchain-handler'
import { LinkProof } from '@ceramicnetwork/blockchain-utils-linking'
import { AccountID } from 'caip'
import * as uint8arrays from 'uint8arrays'
import { verify } from '@stablelib/ed25519'

const verifySignature = async (
  pubKey: Uint8Array,
  message: string,
  signature: Uint8Array
): Promise<Boolean> => {
  const verified = verify(
    pubKey,
    uint8arrays.fromString(message),
    signature,
  )
  return verified
}

const namespace = 'solana'

export async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const pubKey = uint8arrays.fromString(new AccountID(proof.account).address, "base58btc")
  const msg = proof.message
  const sig = uint8arrays.fromString(proof.signature, 'base64')
  const is_sig_valid = await verifySignature(pubKey, msg, sig)
  return is_sig_valid ? proof : null
}

const Handler: BlockchainHandler = {
  namespace,
  validateLink,
}

export default Handler
