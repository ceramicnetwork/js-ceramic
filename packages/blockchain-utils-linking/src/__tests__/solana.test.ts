import { jest } from '@jest/globals'
import { MessageSignerWalletAdapterProps } from '@solana/wallet-adapter-base'
import { SOLANA_MAINNET_CHAIN_REF, SolanaAuthProvider } from '../solana.js'
import { Keypair } from '@solana/web3.js'
import { sign } from '@stablelib/ed25519'
import * as uint8arrays from 'uint8arrays'

const did = 'did:3:bafysdfwefwe'
const privKey =
  'mdqVWeFekT7pqy5T49+tV12jO0m+ESW7ki4zSU9JiCgbL0kJbj5dvQ/PqcDAzZLZqzshVEs01d1KZdmLh4uZIg=='
const chainRef = SOLANA_MAINNET_CHAIN_REF

class MyWalletAdapter implements MessageSignerWalletAdapterProps {
  readonly _keyPair: Keypair

  constructor(keyPair: Keypair) {
    this._keyPair = keyPair
  }

  async signMessage(message: Uint8Array): Promise<Uint8Array> {
    return sign(this._keyPair.secretKey, message)
  }
}

let keyPairEd25519: Keypair

beforeAll(() => {
  keyPairEd25519 = Keypair.fromSecretKey(uint8arrays.fromString(privKey, 'base64'))
  global.Date.now = jest.fn().mockImplementation(() => 666000)
})

afterAll(() => {
  jest.clearAllMocks()
})

describe('Blockchain: Solana', () => {
  describe('createLink', () => {
    test(`create proof for ${chainRef}`, async () => {
      const provider = new MyWalletAdapter(keyPairEd25519)
      const authProvider = new SolanaAuthProvider(
        provider,
        keyPairEd25519.publicKey.toString(),
        chainRef
      )
      const proof = await authProvider.createLink(did)
      expect(proof).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test(`create proof for ${chainRef}`, async () => {
      const provider = new MyWalletAdapter(keyPairEd25519)
      const authProvider = new SolanaAuthProvider(
        provider,
        keyPairEd25519.publicKey.toString(),
        chainRef
      )
      const result = await authProvider.authenticate('msg')
      expect(result).toMatchSnapshot()
    })
  })
})
