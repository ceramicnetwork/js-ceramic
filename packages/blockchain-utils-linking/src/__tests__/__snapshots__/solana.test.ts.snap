// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Solana authenticate create proof for 5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp 1`] = `"0xd042f28e5ff37a2f68f7d7b96adca0d10b5cba3d4d6d056796653479d8a70c0e"`;

exports[`Blockchain: Solana createLink create proof for 5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp 1`] = `
Object {
  "account": "2q7pyhPwAwZ3QMfZrnAbDhnh9mDUqycszcpf86VgQxhF@solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
  "message": "Link this account to your identity

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "KriuHY8KAU87qNeGjdtvutl0TWgSnOPpUDI8fniN8jHGdwcKubYUooJja/hnFZoizq7q66dx5XRPxKS4EcnGDA",
  "timestamp": 666,
  "type": "solana",
  "version": 2,
}
`;
